### PR TITLE
LBAC-38 : Global Property for Every Entity Restriction

### PR DESCRIPTION
**Pull request by VANKINEENITAWRUN**
## Description of what I changed

Added Global property for every entity to customise the Restriction only on few entitites rather than only.

Having global property `locationbasedacess.patientaccess` as `true` implies to have LBAC restriction on it.

## Issue I worked on

see https://issues.openmrs.org/browse/LBAC-38

## Checklist: I completed these to help reviewers :)

- [x] My pull request only contains **ONE single commit**
(the number above, next to the 'Commits' tab is 1).

  No? -> [read here](https://wiki.openmrs.org/display/docs/Pull+Request+Tips) on how to squash multiple commits into one

- [x] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

  No? Unsure? -> [configure your IDE](https://wiki.openmrs.org/display/docs/How-To+Setup+And+Use+Your+IDE), format the code and add the changes with `git add . && git commit --amend`

- [ ] I have **added tests** to cover my changes. (If you refactored
  existing code that was well tested you do not have to add tests)

  No? -> write tests and add them to this commit `git add . && git commit --amend`

- [x] I ran `mvn clean package` right before creating this pull request and
  added all formatting changes to my commit.

  No? -> execute above command

- [x] All new and existing **tests passed**.

  No? -> figure out why and add the fix to your commit. It is your responsibility to make sure your code works.

- [x] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

### CODE REVIEW

File: api/src/main/java/org/openmrs/module/locationbasedaccess/LocationBasedAccessActivator.java [hunk 1]
```diff
@@ -53,9 +53,23 @@ public void started() {
 			    LocationBasedAccessConstants.PERSONATTRIBUTETYPE_UUID);
 	    createGlobalPropertyForLocationAttribute(LocationBasedAccessConstants.REF_APP_LOCATION_USER_PROPERTY_NAME,
 			    LocationBasedAccessConstants.LOCATION_USER_PROPERTY_NAME);
+	    createGlobalPropertyForEntityRestrictions();
 	    log.info("Location Based Access Control Module started");
     }
 
+	private void createGlobalPropertyForEntityRestrictions() {
```
Comment (suthagar23) [inline api/src/main/java/org/openmrs/module/locationbasedaccess/LocationBasedAccessActivator.java]: Can you add comments for this GP, then a developer can simply identify these use cases.

File: api/src/main/java/org/openmrs/module/locationbasedaccess/LocationBasedAccessConstants.java [hunk 2]
```diff
@@ -20,4 +20,11 @@
     public static final String PERSONATTRIBUTETYPE_FORMAT = "java.lang.String";
     public static final String PERSONATTRIBUTETYPE_UUID = "0a93cbc6-5d65-4886-8091-47a25d3df944";
 
+    public static final String PATIENT_RESTRICTION_GLOBAL_PROPERTY_NAME = "locationbasedacess.patientaccess";
```
Comment (suthagar23) [inline api/src/main/java/org/openmrs/module/locationbasedaccess/LocationBasedAccessConstants.java]: what do you think about - `locationbasedaccess.access.patient`? like this for others?

File: api/src/main/java/org/openmrs/module/locationbasedaccess/aop/interceptor/EncounterServiceInterceptorAdvice.java [hunk 3]
```diff
@@ -37,7 +38,8 @@ public Object invoke(MethodInvocation invocation) throws Throwable {
         }
 
         Object object = invocation.proceed();
-        if (Daemon.isDaemonUser(authenticatedUser) || authenticatedUser.isSuperUser()) {
+        String lbacRestriction = Context.getAdministrationService().getGlobalProperty(LocationBasedAccessConstants.ENCOUNTER_RESTRICTION_GLOBAL_PROPERTY_NAME);
+        if (Daemon.isDaemonUser(authenticatedUser) || authenticatedUser.isSuperUser() || !("true".equals(lbacRestriction))) {
```
Comment (suthagar23) [inline api/src/main/java/org/openmrs/module/locationbasedaccess/aop/interceptor/EncounterServiceInterceptorAdvice.java]: the better option would be other way around right? 
`lbacRestriction.equals("true")`

File: api/src/main/java/org/openmrs/module/locationbasedaccess/LocationBasedAccessConstants.java (hunk 2 above)
Comment (VANKINEENITAWRUN) [reply api/src/main/java/org/openmrs/module/locationbasedaccess/LocationBasedAccessConstants.java]: fixed it. please check now

File: api/src/main/java/org/openmrs/module/locationbasedaccess/aop/interceptor/EncounterServiceInterceptorAdvice.java (hunk 3 above)
Comment (VANKINEENITAWRUN) [reply api/src/main/java/org/openmrs/module/locationbasedaccess/aop/interceptor/EncounterServiceInterceptorAdvice.java]: `"true".equals(lbacRestriction))` used it thinking misssing nullpointer exception by using `lbacRestriction.equals("true")` we can change it as lbacRestriction will not be NUll

File: api/src/main/java/org/openmrs/module/locationbasedaccess/LocationBasedAccessActivator.java (hunk 1 above)
Comment (VANKINEENITAWRUN) [reply api/src/main/java/org/openmrs/module/locationbasedaccess/LocationBasedAccessActivator.java]: Added comments.

**Comment by dkayiwa** [comment]
@suthagar23 any more comments for this?

File: api/src/main/java/org/openmrs/module/locationbasedaccess/aop/interceptor/EncounterServiceInterceptorAdvice.java (hunk 3 above)
Comment (suthagar23) [reply api/src/main/java/org/openmrs/module/locationbasedaccess/aop/interceptor/EncounterServiceInterceptorAdvice.java]: If you checking for null pointer expectation, then that needs to be diffrent. 

Here, it's means - you are checking where lbacRestriction is true or false. So the better option would be other way around. Is it?

**Comment by suthagar23** [comment]
@dkayiwa Still some comments need to be addressed!

@VANKINEENITAWRUN Can you check the comments and follow JIRA guidelines also.

File: api/src/main/java/org/openmrs/module/locationbasedaccess/aop/interceptor/EncounterServiceInterceptorAdvice.java (hunk 3 above)
Comment (VANKINEENITAWRUN) [reply api/src/main/java/org/openmrs/module/locationbasedaccess/aop/interceptor/EncounterServiceInterceptorAdvice.java]: okay

**Comment by VANKINEENITAWRUN** [comment]
@suthagar23  following Jira guidelines did not claim the issue as they are not made for ready for work yet. sorry, will make sure everything according to guidelines.

**Comment by dkayiwa** [comment]
@suthagar23 what do you think of this?

**Pull request merged**